utils/rr_graph_walk/rr_graph_walk.py

```diff
@@ -0,0 +1,398 @@
+        """
+
+        def append_edge(edge_dict, key_node, target_node):
+            if key_node not in edge_dict.keys():
+                edge_dict[key_node] = [target_node]
+            else:
+                edge_dict[key_node].append(target_node)
+
+        print("Loading and parsing XML file...")
+
+        # Load and initialize parser
+        parser = ET.iterparse(xml_file, events=("start", "end"))
+        parser = iter(parser)
+
+        # Parse
+        expect_nodes = False
+        expect_edges = False
+
+        for event, element in progressbar(parser):
+
+            # Start
+            if event == "start":
+
+                # Begin "rr_nodes"
+                if element.tag == "rr_nodes":
+                    expect_nodes = True
+                # Begin "rr_edges"
+                if element.tag == "rr_edges":
+                    expect_edges = True
+
+            # End
+            elif event == "end":
+
+                # End "rr_nodes"
+                if element.tag == "rr_nodes":
+                    expect_nodes = False
+                # End "rr_edges"
+                if element.tag == "rr_edges":
+                    expect_edges = False
+
+                # Got a complete node element
+                if expect_nodes and element.tag == "node":
+
+                    # Find the "loc" tag
+                    xml_loc = element.find("loc")
+
+                    # Append the node
+                    node = RoutingGraph.Node(
+                        id=int(element.get("id")),
+                        type=element.get("type"),
+                        xlow=int(xml_loc.get("xlow")),
+                        ylow=int(xml_loc.get("ylow")),
+                        xhigh=int(xml_loc.get("xhigh")),
+                        yhigh=int(xml_loc.get("yhigh"))
+                    )
+                    self.nodes[node.id] = node
+
+                # Got a complete edge element
+                if expect_edges and element.tag == "edge":
+                    src_node = int(element.get("src_node"))
+                    dst_node = int(element.get("sink_node"))
+
+                    # Append edge
+                    append_edge(self.edges_from, src_node, dst_node)
+                    append_edge(self.edges_to, dst_node, src_node)
+
+                # Clear the element
+                if element.tag != "loc":
+                    element.clear()
+
+        # Clean up
+        del parser
+        gc.collect()
+
+        print(
+            "{} nodes, {} edges".format(len(self.nodes), len(self.edges_from))
+        )
+
+    def node_to_string(self, node_id):
+        """
+        Converts node information to string
+        """
+
+        node = self.nodes[node_id]
+
+        return "%s:%d [%d,%d,%d,%d]" % (
+            node.type, node.id, node.xlow, node.ylow, node.xhigh, node.yhigh
+        )
+
+    def verify_route(self, route, walk_direction):
+        """
+        Checks wheter a given route is valid
```

Comment:
For the docstrings, document what the arguments are and their purpose.

Reply:
Done.